### PR TITLE
[skin][estuary] Fix visibility condition on topbaroverlay

### LINKED ISSUE #20078: The timeline isn't displayed when pausing a video with JSON-RPC
### Describe the bug
Here is a clear and concise description of what the problem is:

The timeline isn't displayed when pausing a video with [JSON-RPC](https://kodi.wiki/view/JSON-RPC_API/v12#Player.PlayPause). The timeline is displayed when I pause with my TV remote (and HDMI-CEC):
pause

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

When you pause with JSON-RPC, the timeline should be displayed.

## Actual Behavior

The timeline isn't displayed.

### To Reproduce

1. _Allow remote control via HTTP_ (_Settings_ / _Services_ / _Control_).
2. Play a video.
3. Execute (by changing the IP address) : `curl -d '{ "jsonrpc": "2.0", "method": "Player.PlayPause", "params": { "playerid": 1 }, "id": 1 }' -H 'Content-type: application/json' -X POST 127.0.0.1:8080/jsonrpc`
4. The timeline isn't displayed.

### Debuglog

The debuglog can be found here:
```
2021-08-24 18:46:41.375 T:8794    DEBUG <CWebserver[8080]>: request received for /jsonrpc
2021-08-24 18:46:41.379 T:480     DEBUG <general>: ------ Window Init (DialogSeekBar.xml) ------
2021-08-24 18:46:41.392 T:8789    DEBUG <general>: CDVDAudio::Pause - pausing audio stream
2021-08-24 18:46:41.394 T:480     DEBUG <general>: ------ Window Init (Custom_1109_TopBarOverlay.xml) ------
```

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

I didn't have this problem with Kodi 18.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: OSMC (Open Source Media Center) 2021.08-1 (kernel: Linux 5.10.32-1-osmc) 
 - Kodi version:
   - Build: 19.1 (19.1.0)
   - Git: 20210715-osmc
   - Compiled: 2021-07-15
- Skin: Estuary

The bug is also with Kodi 20.0-ALPHA1 (19.90.101) on Ubuntu 20.04.3 LTS.

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by enen92**
## Description
OSD redesign in Matrix introduced a new top bar overlay which is automatically closed once the user gives some input to kodi as a result of some action (e.g. seek). However this has introduced some sort of regression for the pause state.
If the user executes play/pause via the GUI (by hitting the button) the overlay will stay displayed on top of video (since the idle timer is restarted). However if the user does the same action via JSON using for example:

```
http://localhost:8152/jsonrpc?request={"jsonrpc":"2.0","method":"Player.PlayPause","params":{"playerid":1},"id":8}
```
Sometimes the overlay is not show or ir autohides too quickly. This happens because those methods don't reset the idle timer (which is correct as they are not input methods). Since this seems to be something wanted by remote developers (and users) and we have differences of behaviour for the same action, lets keep it open for Player.Paused.

## Motivation and context
Fix https://github.com/xbmc/xbmc/issues/20078

## How has this been tested?
Runtime tested

## What is the effect on users?
Paused state will be shown either if paused from the GUI or from JSON, estuary keeps it open if `Player.Paused` is true

## Screenshots (if appropriate):

image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by enen92** [comment]
This ends up modifying the current default behavior (the seekbar doesn't autohide if paused for more than 5 seconds) that I prefer as a user.
Maybe we can have a skin setting to keep it open if the user desires so that the current default behavior is not changed?

**Review by jjd-uk — COMMENTED** [review]
The auto-hide behaviour was a specific feature of the new top seekbar as it had been requested by users see https://github.com/xbmc/xbmc/pull/18448 so this probably shouldn't be removed.

Besides the skin code change here makes no sense to me., the whole point of L12 is to time out the seekbar when paused, if you want to stop this behaviour then the cleanest solution is to remove L12 entirely.

**Comment by jjd-uk** [comment]
> This happens because those methods don't reset the idle timer (which is correct as they are not input methods)

From my completely uneducated in the C++ side of Kodi view, if remote json commands like this replicate direct local input methods they should be treated the same as using a local input device (keyboard/mouse) and so should reset the idle timer.

**Comment by enen92** [comment]
I think it comes down to the fact the Player json methods are not Input methods - they affect the Player state only. Personally I think this is right, json methods should have headless in mind, avoid dependencies on GUI, input, etc.
The estuary top seekbar depends on the idle timer, that's intrinsically linked to input. Still, kodi json-rpc has "raw" input method support via `Input.ExecuteAction`. For json based remotes to mimic the physical remote behaviour, those methods should be used instead of the player state methods. It's not only the idletimer that's affected, the GUI sound of keypresses are also not executed if not triggered via Input.
The only point I have doubts is in a (far far) distant future when and if Kodi is able to run headless. Will input methods make sense in that scenario? Regardless, too far to a realistic possibility so I'd say the best and safest option for the near future is to use Input via json.

TLDR: I still don't think there's anything to fix on the kodi core side, it comes down to the json clients of Kodi to choose the appropriate methods to execute.

@Tolriq @SyncedSynapse your input is appreciated.

**Comment by Tolriq** [comment]
This is something I fought for and someone no more here did not agree and hated me for that ;)

>  If I change player state via remote command then surely system hasn't remained idle as I've made a user initiated action???? so idle timer should be reset.

Is IMO correct, press right on a physical remote or keyboard or a phone should trigger the same behavior. Same for play / pause and all the rest.

Remotes have no way to detect all the possible cases nor access the keymaps to be able to reproduce what the user have configured on Kodi side. (Keymaps).

In the end Inputs are mapped to commands. Json gives direct access to commands but no access to keymaps. This would add another layer of translation that would give strange results as different from Kodi versions to versions, but still not matching what the user could have manually forced via keymaps.

About headless or not, currently Kodi is not headless so the actions should do the same, if one day Kodi is headless then there's still no issue as idle times for the GUI or the screensaver kicking it becomes irrelevant.

> For json based remotes to mimic the physical remote behavior, those methods should be used instead of the player state methods

Do not agree, specially because of the keymaps file that remote have no access to, see above.

Another example: Use Json to play a media. Screensaver is visible, since it's not an input the screensaver stays over the media? Does not really make sense (But was the case for a long time, it's now fixed)

**Comment by enen92** [comment]
> About headless or not, currently Kodi is not headless so the actions should do the same, if one day Kodi is headless then there's still no issue as idle times for the GUI or the screensaver kicking it becomes irrelevant.
> 
> > For json based remotes to mimic the physical remote behavior, those methods should be used instead of the player state methods
> 
> Do not agree, specially because of the keymaps file that remote have no access to, see above.
> 

Thanks for the input. Regarding this point I don't think I follow your logic. Why would you need access to keymaps via json in this particular situation? I don't think this is an irrefutable truth: it might be useful in other use cases but I fail to see the need here.
"We" are not mapping keys to actions, we are linking app buttons to specific actions. As far as I understood, kodi maps key input (mouse, remote, joystick, etc) to high level actions. Some of those actions are in fact `input` based-actions (e.g. some of the default ones for the return key, left mouse click, etc).
In this particular situation, you have a player osd on your remote app and a player osd on kodi. I'd say both should have the same end behaviour (i.e. the end ACTIONS) on the kodi GUI. The problem is that when you use the GUI you use the Input subsystem to end up executing the PlayPause action (because it is defined on the `onclick` tag of the play/pause button in estuary). Via json if using Player.Playpause you bypass the input layer and execute the action directly. Doing a "non-sense analogy" you don't have a `onclick` nor an input system.
To achieve the exact same behaviour the PLAY_PAUSE action should be wrapped by the the Input method instead when executed from Json.

OFC you could argue that if a user has modified the keymap (e.g. modify the "return" key to execute some other action rather than "enter") the behaviour when cliking the play/pause osd button would then differ from the one you execute from the remote app (if the user manages to show the OSD via some other key rather than enter :) ). But up to what point is this relevant? A remote application is not a kodi peripheral so it doesn't need keymaps. You could even hypothetically achieve the 1:1 clone with (far too much) work on your remote by creating some sort of keymap subsystem on the remote itself. So that when touching a button on your app it could execute the same action (via JSON input.sendaction) that kodi gui elements execute. There you'd need at least read access to the touch keymap in yhe app. But...taking into consideration the amount of work involved (from both sides) and the (really really) edge case, is it worth it to even think about it?

**Comment by enen92** [comment]
> Another example: Use Json to play a media. Screensaver is visible, since it's not an input the screensaver stays over the media? Does not really make sense (But was the case for a long time, it's now fixed)

I see this as an unrelated comparison btw, the player itself should wake the screensaver and reset dpms (if the player is not part of a screensaver itself) and has little (if none) relation to input

**Comment by Tolriq** [comment]
You are mixing things with just this play pause issue and the global way Kodi works with keymaps.

Another simple example: Arrows during playback, can be skip, adaptative seek, change chapters, can be changed via keymaps. (Different behaviors depending on Kodi version). Yet it's impossible to reproduce the same behavior from remotes and we need to have different actions depending on Kodi version hoping it will mostly match. (Or we can use EventServer that allows proper integration of the keymaps, but as discussed many time this is a way less secure protocol).

If users configure Kodi to have the arrow on it's physical remote to do something it expect the phone arrow to do the same thing, it's IMO and from 10 years of users feedback insane to ask them to redo the configuration on the remote side too. (Not talking about the dozens of missing commands from JSON making this virtually impossible)

And there's not just the GUI that handle the input system. Many physical remotes have a play button mapped to the play/pause command. Why a pause button on a physical remote would act differently than a pause button on phone?

Another example, pressing space on the keyboard does pause the media and show Kodi GUI paused bar. Pressing pause on the remote pause the media and show Kodi GUI. But doing so via JSON should not show the GUI paused bar then since it's not input?
Same the paused bar auto hide when unpausing from all other input systems but not from JSON?
What about EventServer? This is another way to send commands but those are input so react also differently.

Anyway this was discussed for years now, you asked my feedback I gave it (and it include feedback from many many users over the last 10 years).

API and actions should be consistent over everything, GUI, Python, JSON, Event Server, Inputs, .... currently each behave differently and it's a mess for the users, for the devs and in the end for everyone.
There's history and everything that explains the current situation, but no the current situation is not the goal and the proper thing, it's just what it is now, and it should not serve as a reason to not fix things when it's possible.

> I see this as an unrelated comparison btw, the player itself should wake the screensaver and reset dpms (if the player is not part of a screensaver itself) and has little (if none) relation to input

I used that as an explanation for a very long lasting issue on Kodi about the difference between the internals in Kodi ;)
But you can use the same thing with other commands. Sending a command to go to the home screen of Kodi from JSON should also stop the screensaver? Or it's not an Input command and it should not? Or it's not an input command but a GUI command so we make some exception? But then why some exception and some not? How the users know what will do what?

> Via json if using Player.Playpause you bypass the input layer and execute the action directly. Doing a "non-sense analogy" you don't have a onclick nor an input system.
To achieve the exact same behaviour the PLAY_PAUSE action should be wrapped by the the Input method instead when executed from Json.

Again IMO and from experience, this is the opposite, what about EventServer? What about Python? 
Imagine a python addon that listen to some home automation event and pause the media, why should it behave differently?
Why should the python addon have to wrap anything to an input system (that could be broken via keymaps) to achieve something as simple as pausing a media.

The proper answer is that everything should be the same whatever the system. All the entry points call the same end point that trigger all the necessary things like resetting idle and everything.

Anyway as said just my opinion and experience.

**Comment by jjd-uk** [comment]
If we wanted full consistancy then I'd say there ideally needs to be a json input api that does through the input layer, then there could be a json.xml for default mapping in the same way as other input devices which could then be altered. But I would guess this is a very big refactor job and thus unlikely to change, unless it would make a good GSOC project.

Without any big refector happening then I think @enen92 is saying the best way to ensure some consistancy with a physical remote is by using `Input.ExecuteAction` because as least that does go through the input layer.

**Comment by Tolriq** [comment]
> If we wanted full consistency then I'd say there ideally needs to be a json input api that does through the input layer, then there could be a json.xml for default mapping in the same way as other input devices which could then be altered. 

Not really, there's 2 different things: Input that can be mapped. No need for a json.xml, as with EventServer we just need to tell what mapping we want to use, keyboard, remote, xbox, ... No need to have the user duplicate their configuration.
For the record this was added relatively recently as ButtonEvent and I'm glad it was not refused despite being refused as an idea a lot times before.

And then there's direct commands like play / pause / ... those should do the command and should not be possible to map, but they should behave internally in Kodi in the same ways as the others. They are events too as any other input system they should reset the idle.

It's the same issue when playing stuff, depending on how you start a media, from Json, from EventServer, from Python or from the GUI it will behave differently and will have many subtle differences. The input layer is not the place to fix those kind of things, the execution layer should be.

Inputs -> Mapping -> Converter -> Command -> CommandHandler
GUI -> Command -> Converter -> CommandHandler
Python -> Command -> Converter -> CommandHandler
Json -> Command -> Converter -> CommandHandler

Currently there's no converter and single CommandHandler, so each part on a lot of things have code to directly do things and each have small differences and more maintenance cost.

Furthermore again: `Input.ExecuteAction` does not cover everything, any action with parameters can't be used from Json

_______
As a resume: 

Json offer:
- Direct play pause action do things but not event but have useful return value and everything to offer a proper experience to users.
- Input.ExecuteAction action do things directly but event (Useful for some commands but not really for those who have a dedicated API)
- Input.ButtonEvent action do things via keymaps and is event (Allows remote to mimic physical remotes)

There is absolutely no reason for remote makes to cripple their application with the very limited Input.ExecuteAction for everything that have a real JSON Api.

An input action is an input action, the fact that Kodi have something called Input does not make the other entry points not input actions. And the fact that something is currently only working for Inputs does not make it less worthy for the other entry points.

Using the wrong API to trick the software internals is never the solution, it's a workaround and will always be.

Again why would a python addon that react to someone ringing at the door to pause a media, should not use the pause command but cheat to make itself pass as an input system....

**Comment by enen92** [comment]
@Tolriq thanks much for your provided input here. I cannot say I agree with you on all the points but I respect your opinion and experience on the matter. It will surely help the project on the long run as the information provided will stay in memory.

Regarding the issue at hand here, I thought about it from a different angle and I'm proposing https://github.com/xbmc/xbmc/pull/20816 as a possible fix. I think it highly depends at the angle we observe the root cause of the problem :)

Thanks again
